upgrade to Julia 1.x

johnnychen94 opened the pull request:
The update work is split into several PRs(e.g., #27, #31) and will be merged into this PR/branch first. There will be a lot of changes, so it's not suitable to directly review this.

Basically, the roads are:

- [x] fix 0.7 deprecations (#27)
- [x] fix 1.x deprecations (#31)
- [x] bug: #38 
- [x] bug: #39
- [ ] upgrade to OffsetArray v1.0
- [ ] update docs

When all things get settled done, I will merge this branch into master.

Evizero committed:
update code to 0.7 (#27)

change indexes to indices

change Range to AbstractRange

use Base.showarg

update CI and README

fix basic deprecations

fix showcompact deprecation

fix channel ops and tests

fix convert ops

work on mapfun ops

johnnychen94 committed:
update codebase to 1.0 and 1.x (continue #27) (#31)

* add Project.toml

* update test for cache and minor fix

* update test for rotation and minor fix

* update flipdm to reverse

https://github.com/JuliaLang/julia/pull/26369

* update test for crop and minor fix

* update test for resize

* update test for scale and minor fix

* update test for zoom and minor fix

* WIP: update test for distortion and minor fix

* update test for either and minor fix

* fix pipeline test

* update summary reference for distortedview

* WIP: fix tests for augment and operations

* drop julia 0.7 and update CI

* use FileIO v1.1.0

https://github.com/JuliaIO/FileIO.jl/issues/245

* add 1.x tests to CI

* revert back unnecessary changes

* test windows in travis

* fix randomly failed test cases for inacuraccy reasons

* update to julia 1.1

* maybe_copy doesn't explain its usage
* replace Slice by OffsetArrays.IdentityUnitRange

https://github.com/JuliaArrays/OffsetArrays.jl/pull/62

* update to julia 1.2 -- part I

* WIP: update to julia 1.2 -- part II

detect some type instable test cases, only the outmost wrapper is
instable here

* unify how types are displayed

Only meta operations such as Either, Pipeline doesn't have Agumentor
prefix

* add more test versions

* move safe_rand to compat.jl

rand is thread-safe in julia 1.3

* don't allow failures for julia 1.3

* fix commit "WIP: update 1.0"

* use explicit and intuitive CartesianIndex(1, 1)

* add method specialization for tweight

* allow FileIO 1.2

* fix type instability for Either

* update reference for FixedPointNumbers v0.7

* try: relax equality check for scale

* restore test cases in tst_scale.jl

* Revert "add method specialization for tweight"

This reverts commit a6983828a9a7ac581b6f314db8a31b7188c26656.

* Augmentor v0.6.0-pre

johnnychen94 committed:
add CompatHelper github action

johnnychen94 committed:
switch to github action for CI

Github action supports more concurrent jobs than travis CI

johnnychen94 committed:
disable travis CI in favor of github action

johnnychen94 commented:
Commits 40820af and 03d97a0 were pushed to master and rebased to this branch; then I realized that they won't actually work before this PR is ready. To disable daily notification of failed GitHub action jobs, I deleted these two commits from the master branch so that there will be no scheduled jobs running.

To keep track of history, this PR shouldn't be merged with "squash and merge" option.

Evizero commented on src/operations/distortion.jl:
````diff
@@ -93,7 +93,7 @@ augment(img, ElasticDistortion(15, 15; scale = 0.1))
 augment(img, ElasticDistortion(10, 10; sigma = 4, iter=3, border=true))
 ```
-""" ->
+"""
````
are you sure?

johnnychen94 replied:
Ah, these two changes are made in 8dbe2b1 (#27 ) and my changes #31 is built on top of that.

I haven't observe such `->` usage after Julia 1.0; so I guess it's an deprecated syntax

```julia
help?> ElasticDistortion
```
works normally.

Evizero commented on test/operations/tst_mapfun.jl:
```diff
@@ -101,13 +103,14 @@ end
             img = OffsetArray(rgb_rect, -2, -1)
-            res = @inferred(Augmentor.applyeager(AggregateThenMapFun(mean, (x,a)->x-a), img))
+            @test_broken false # FIXME: type stability below from "mean"
+            res = (Augmentor.applyeager(AggregateThenMapFun(mean, (x,a)->x-a), img))
```
Out of interest: Any idea where this issue came from? `mean` itself?

johnnychen94 replied:
This test isn't broken anymore.

johnnychen94 committed:
specialize tweight on FixedPoint conditionally (#42)

If upstream package (e.g., Interpolations, FixedPointNumbers) doesn't
provide such specialization, then we add one here.

This trick avoids incremental compilation warnings

johnnychen94 committed:
relax accuracy requirement for random failed tests

OffsetArray might hit accuracy issue for affine
Some affine operations (e.g., scale and zoom) might hit accuracy
issues when images are upsampled

This is a temporarily workaround to issue #38 (by disabling the tests)

johnnychen94 committed:
put back inferred macro for previous broken tests

johnnychen94 committed:
stop testing julia 1.1 and 1.2 in CI

johnnychen94 commented:
Since this branch is functioning now with all deprecations fixed, I'm merging this branch to master. From now on, compat upgrades(e.g., OffsetArrays 1.0 is a breaking change) and new features can be added to the master branch separately.

johnnychen94 commented:
@Evizero since most CI are moved to github action, can you uninstall appveyor?

Submitting coverage report via github action doesn't work for PRs from other forks. (It's an upstream github action issue), hence travis is still required to upload coverage report

Evizero commented:
done
